### ISSUE TITLE
Keras

### ISSUE THREAD

**Issue by junxnone** [junxnone/tio#9]
Keras is a high-level API to build and train deep learning models.

# Reference
- [ ] [keras github](https://github.com/keras-team/keras)
- [ ] [Tesnsorflow keras guide](https://www.tensorflow.org/guide/keras)
- [ ] [Keras en doc](https://keras.io/)
- [ ] [Keras zh-cn doc](https://keras.io/zh/)
- [ ] [Keras zh-cn doc source](https://github.com/keras-team/keras-docs-zh)
- [ ] [keras core layer](https://keras-cn.readthedocs.io/en/latest/layers/core_layer/)
- [ ] [文本与序列预处理](https://blog.csdn.net/zzulp/article/details/76146947)
- [ ] [用Keras和直方图均衡化---数据增强](https://cloud.tencent.com/developer/article/1052765)

# Brief
- [ ] Input
- [ ] [图像预处理](https://keras.io/zh/preprocessing/image/)
- [ ] Layers
- [ ] model
- [ ] Optimizers
- [ ] Losses
- [ ] Metrics
- [ ] Activations

# 后端 Backend
Keras 有三个后端实现可用:
- TensorFlow 是由 Google 开发的一个开源符号级张量操作框架。
- Theano 是由蒙特利尔大学的 LISA Lab 开发的一个开源符号级张量操作框架。
- CNTK 是由微软开发的一个深度学习开源工具包。

### 全局切换后端
```
$ HOME/.keras/keras.json
```
```
{
    "image_data_format": "channels_last",
    "epsilon": 1e-07,
    "floatx": "float32",
    "backend": "tensorflow"
}
```

### 暂时切换后端

定义环境变量 KERAS_BACKEND，这会覆盖配置文件中定义的内容：
```
KERAS_BACKEND=tensorflow python -c "from keras import backend"
Using TensorFlow backend.
```

**Comment by junxnone** [comment]
# model
##  1. the Sequential model
### [Sequential 模型 API](https://keras.io/zh/models/sequential/#sequential-api)
- [ ] [fit()](https://keras.io/zh/models/sequential/#fit)

##  2. the Model class used with the functional API.

## 3. fully-customizable models

通过继承 Model 类并在 call 方法中实现你自己的前向传播，以创建你自己的完全定制化的模型.
在 call 中，你可以指定自定义的损失函数，通过调用 self.add_loss(loss_tensor)

**Comment by junxnone** [comment]
# Optimizers

**Comment by junxnone** [comment]
# Metrics
评价函数用于评估当前训练模型的性能。当模型编译后（compile），评价函数应该作为 metrics 的参数来输入。评价函数和 损失函数 相似，只不过评价函数的结果不会用于训练过程中。
**参数**
- y_true: 真实标签，Theano/Tensorflow 张量。
- y_pred: 预测值。和 y_true 相同尺寸的 Theano/TensorFlow 张量。
## 评价函数
- binary_accuracy
categorical_accuracy

**Comment by junxnone** [comment]
#  Activations
激活函数可以通过设置单独的激活层实现，也可以在构造层对象时通过传递activation参数实现

## 预定义激活函数
- [ ] softmax
- [ ] elu
- [ ] selu
- [ ] softplus
- [ ] softsign
- [ ] relu
- [ ] tanh
- [ ] sigmoid
- [ ] hard_sigmoid
- [ ] linear
- [ ] [高级激活函数](https://keras.io/zh/layers/advanced-activations/)

**Comment by junxnone** [comment]
# 后端 Backend
Keras 有三个后端实现可用:
- TensorFlow 是由 Google 开发的一个开源符号级张量操作框架。
- Theano 是由蒙特利尔大学的 LISA Lab 开发的一个开源符号级张量操作框架。
- CNTK 是由微软开发的一个深度学习开源工具包。

## 如何切换后端
### 全局切换后端
```
$ HOME/.keras/keras.json
```
```
{
    "image_data_format": "channels_last",
    "epsilon": 1e-07,
    "floatx": "float32",
    "backend": "tensorflow"
}
```

### 暂时切换后端

定义环境变量 KERAS_BACKEND，这会覆盖配置文件中定义的内容：
```
KERAS_BACKEND=tensorflow python -c "from keras import backend"
Using TensorFlow backend.
```

**Comment by junxnone** [comment]
# keras 构建 model 的方式：
- [ ] model = Sequential()
- [ ] model = Model(inputs=inputs, outputs=predictions)
- [ ] model = Model.from_config(config)
- [ ] model = model_from_json(json_string)
- [ ] model = model_from_yaml(yaml_string)
- [ ] 继承 Model 类
- [ ] 重用训练好的模型
